### PR TITLE
25 fix project env

### LINKED ISSUE #25: `@install_project` decorator does not work for step explicitly decorated with `@batch`
`argv`  on the batch machine for a step with  `@batch` does not contain `--with batch` or similar, thus the `is_task_local` function incorrectly returns False.

### PR DESCRIPTION
**Pull request by bishax**
Closes #25

This long-running PR started as a straight-forward fix that emerged more and more problems as time came on which has resulted in a drastic refactoring and rescoping of the core functionality of this library.

There is a companion PR (#28) that goes even further based on the fact that Metaflow `2.5.2` now follows symlinks (with cycle detection) and that the metaflow extensions API has changed in a big way.
This PR almost keeps feature parity; however #28 makes vast simplifications in exchange for a small loss of functionality.

---

**General**

- Simply GH actions - use AWS secrets manager; remove uneeded installs
- DOCUMENTATION - README.md provides a reasonable starting point by use
- Do not modify default --package-suffixes - let users decide and avoid
  potential security issues by unexpectedly sharing sensitive config
- Pin to metaflow<2.5 due to breaking changes in the extensions API
- Tidy-up tests; better matching them to behaviours under test; and
  changing them based on changed behaviour
- Fixed broken utils for detecting whether a task is local or not

**@pip**

- Added `safe` boolean toggle - provides speed in exchange for safety
  during development.

**Project environment**

- `project` environment is no longer set to default
- No longer install a local package automatically - @pip should be used
  instead if needed
- Export $MFPYTHON bash variable for using the right Python executable
  in preinstall scripts
- Preinstall packages feature removed - this can be done using `
  $MFPYTHON` in a `preinstall-*.sh` script (will not work locally on
  Conda! - Too complex and error-prone)
- [Based on interim changes in this PR that are currently used in OJO] Job package is no longer added to `pkg_self/` - this means we don't
  need to worry about `PYTHONPATH` or `@pip` having inconsistent
  behaviour locally vs remote BUT we do have to run locally from the
  project-level directory and be careful that there are no filename
  collisions.

---

**Old PR description here for posterity**
<details>

`@install_project` decorator is removed and instead the project package is installed either:
1) **Batch (inc. Batch + Conda):** In bootstrap commands
2) **Conda (local):** As before, in `prepare_step_wrapper`

Broken general utils for detecting whether a task is local are removed.
For `ProjectEnvironment` it should test whether the given step *will be* executed in Batch, not whether that code is currently running on a batch machine - we do this by inspecting the decorators that get added to the flow object. This is a more robust check than `platform_arch_mismatch` , not falling prey to doubly installing on Linux.
For `PipStepDecorator` it takes  `metaflow`s install path being `"/metaflow/metaflow"` as the sign that we are executing on Batch. 

Incidental bugs uncovered and fixed:
- If project has a top-level `requirements.txt` this gets over-written if there is a `requirements.txt` alongside the flow!!! (or any similarly conflicting filenames!)
  - **Solution:** put the local package in `pkg_self/` to avoid conflict (pip install still makes these files accessible)
- If using a Conda environment without the project environment then `ProjectEnvironment` bootstrap commands are still added
  - **Solution:** check `DEFAULT_ENVIRONMENT` and only add those commands if it is `"project"` 

You may want to dismiss: Tests refactored to allow running some AWS locally (Mark `daps-utils` tests `ci_only`).

</details>

### CODE REVIEW

File: metaflow_extensions/plugins/project_environment.py [hunk 1]
```diff
@@ -94,15 +106,12 @@ def wrapped_prepare_step_environment(self, *args, **kwargs):
...
-        # match the local arch (e.g. MacOS). In the "edge case" where the local
-        # arch and batch arch match (i.e. are both Linux) then no harm
-        # will be done by doubly installing the same packages, both here and
-        # with project_environment.bootstrap_wrapper
-        if platform_arch_mismatch(env_id):
+        if will_task_be_batch(args[0], self.flow):
             return env_id
 
+        if platform_arch_mismatch(env_id):
+            raise AssertionError("Shouldn't be happening!")
```
Comment (jaklinger) [inline metaflow_extensions/plugins/project_environment.py:113]:
```suggestion
        if platform_arch_mismatch(env_id):
            raise AssertionError("Platform architecture mismatch detected for this step. This shouldn't happen - contact your metaflow_extensions developer.")
```

**Review by jaklinger — CHANGES_REQUESTED** [review]
Looks good from review - just some minor comments. Before approving I'll run this alongside https://github.com/nestauk/ojd_daps/pull/329 to ensure OJO integration

File: metaflow_extensions/plugins/project_environment.py [hunk 2]
```diff
@@ -205,13 +205,12 @@ def bootstrap_commands(self, step_name):
...
         for pkg_spec in filter(None, map(get_pkg_spec, PREINSTALL_PKGS)):
             cmds += [
-                f"{pip_install} --upgrade pip",
                 f"{pip_install} {pkg_spec} 1> /dev/null",
             ]
 
         # Install flow project
-        cmds.append(f"{pip_install} pkg_self/.")
-
+        cmds.append(f"{pip_install} -e pkg_self/. || export PYTHONPATH=$PWD/pkg_self")
```
Comment (jaklinger) [inline metaflow_extensions/plugins/project_environment.py:215]: @bishax this took me about a week to fix in production... there were two separate situations that I was in:

1. Running metaflow (with batch steps) locally, which required the `-e` as discussed earlier (to allow for non-client packages that aren't really meant to be installed, such as a daps)
2. Running metaflow from an `amazonlinux2` container. For some reason, batch steps launched from a container like this (locally, and also in production) have problems running the line `pip install pkg_self/.`. No idea why, and I tried *everything*. In the end I can only recommend the nuclear option of not installing the package at all, and instead just exporting the PYTHONPATH. Very shit option!

For info, the error message I get just says that `.egg_info` can't be found. All threads / docs / q&a online seem to suggest that it's related to updating `setuptools`, but I had previously tried that and other debugging, even breaking down the install, all to no avail.

Comment (bishax) [reply metaflow_extensions/plugins/project_environment.py:215]: :slightly_frowning_face: Sorry it took so long, sounds like a right pain in the arse!

Do you have any of those links still around? And to confirm when you say "locally" you mean that running an `amazonlinux2` container on your laptop reproduces the issue?

Comment (jaklinger) [reply metaflow_extensions/plugins/project_environment.py:215]: The "full" error message is simply:

```
ERROR: No .egg-info directory found in /tmp/pip-pip-egg-info-f55pfh_d'
```

and the rabbit hole of links start from:

* https://github.com/actions/setup-python/issues/165
* https://github.com/pypa/pip/issues/8760
* https://stackoverflow.com/questions/26091641/what-does-a-could-not-find-egg-info-directory-in-install-record-from-pip-mean
* https://setuptools.pypa.io/en/latest/deprecated/python_eggs.html

> And to confirm when you say "locally" you mean that running an amazonlinux2 container on your laptop reproduces the issue?

Yes that's right:

https://github.com/nestauk/ojd_daps/blob/dev/ojd_daps/config/metaflowtask/Dockerfile-base
https://github.com/nestauk/ojd_daps/blob/dev/ojd_daps/config/metaflowtask/Dockerfile

**Pull request merged**